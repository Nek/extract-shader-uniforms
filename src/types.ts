const Scalar = {
    bool: "bool",
    int: "int",
    uint: "uint",
    float: "float",
} as const;

const Vector = {
    vec2: "vec2",
    vec3: "vec3",
    vec4: "vec4",
    bvec2: "bvec2",
    bvec3: "bvec3",
    bvec4: "bvec4",
    ivec2: "ivec2",
    ivec3: "ivec3",
    ivec4: "ivec4",
    uvec2: "uvec2",
    uvec3: "uvec3",
    uvec4: "uvec4",
} as const;

const Matrix = {
    mat2: "mat2",
    mat3: "mat3",
    mat4: "mat4",
    mat2x2: "mat2x2",
    mat2x3: "mat2x3",
    mat2x4: "mat2x4",
    mat3x2: "mat3x2",
    mat3x3: "mat3x3",
    mat3x4: "mat3x4",
    mat4x2: "mat4x2",
    mat4x3: "mat4x3",
    mat4x4: "mat4x4",
} as const;

const Sampler = {
    sampler2D: "sampler2D",
    sampler3D: "sampler3D",
    samplerCube: "samplerCube",
    sampler2DArray: "sampler2DArray",
    samplerCubeShadow: "samplerCubeShadow",
    sampler2DShadow: "sampler2DShadow",
    sampler2DArrayShadow: "sampler2DArrayShadow",
    isampler2D: "isampler2D",
    isampler3D: "isampler3D",
    isamplerCube: "isamplerCube",
    isampler2DArray: "isampler2DArray",
    usampler2D: "usampler2D",
    usampler3D: "usampler3D",
    usamplerCube: "usamplerCube",
    usampler2DArray: "usampler2DArray",
} as const;

export namespace GL {
    enum SplDimension {
        _2D = "2D",
        _3D = "3D",
        _Cube = "Cube",
        _2DArray = "2DArray",
        _CubeShadow = "CubeShadow",
        _2DShadow = "2DShadow",
        _2DArrayShadow = "2DArrayShadow",
    }

    enum Scl {
        Bool = "bool",
        Int = "int",
        Uint = "uint",
        Float = "float",
    }

    type Pre<T extends Scl> = T extends Scl.Bool ? "b" : T extends Scl.Int ? "i" : T extends Scl.Uint ? "u" : "";

    type Spl<T extends Scl, D extends SplDimension> = T extends Scl.Float
        ? `sampler${D}`
        : T extends Scl.Int | Scl.Uint
          ? D extends SplDimension._CubeShadow | SplDimension._2DShadow | SplDimension._2DArrayShadow
              ? never
              : `${Pre<T>}sampler${D}`
          : never;

    type Vec<T extends Scl, N extends 2 | 3 | 4> = `${Pre<T>}vec${N}`;
    type MatCols = 2 | 3 | 4;
    type MatRows = 2 | 3 | 4 | undefined;
    type Mat<COL extends MatCols, ROW extends MatRows> = ROW extends undefined ? `mat${COL}` : `mat${COL}x${ROW}`;

    type Prp = Emb | Arr<Emb, number, number | undefined>;

    type StcDef = {
        [propName: string]: Prp;
    };
    type StcName = string & { __isStcName: true };
    type StcsDict = Record<StcName, StcDef>;

    type Emb = StcName | Scl | Spl<Scl, SplDimension> | Vec<Scl, 2 | 3 | 4> | Mat<MatCols, MatRows>;

    type Arr<T extends Emb, X extends number, Y extends number | undefined> = (Y extends undefined
        ? `[${X}]`
        : `[${X}, ${Y}]`) & { __isArr: true; __type: T };

    export namespace Uniform {
        export namespace Data {
            export const Basic = {
                ...Scalar,
                ...Vector,
                ...Matrix,
                ...Sampler,
            } as const;

            export type Basic = keyof typeof Basic;

            export function isBasicData(type: any): type is Basic {
                return Object.values(Basic).includes(type as Basic);
            }

            export function inferBasicData(type: any): Basic {
                if (isBasicData(type)) {
                    return type;
                }
                throw new Error(`Invalid type: ${type}`);
            }

            export interface Scope {
                [key: string]: keyof typeof Basic | Scope | Array1<keyof typeof Basic | Scope, number> | Array2<keyof typeof Basic | Scope, number, number>;
            }

            export type Array1<T extends keyof typeof Basic | Scope, N extends number> = {
                [index: number]: T;
                length: N;
            };

            export type Array2<T extends keyof typeof Basic | Scope, N extends number, M extends number> = {
                [index: number]: Array1<T, M>;
                length: N;
            };

            export const Setters = {
                [Basic.bool]: "uniform1i",
                [Basic.int]: "uniform1i",
                [Basic.uint]: "uniform1ui",
                [Basic.float]: "uniform1f",
                [Basic.vec2]: "uniform2f",
                [Basic.vec3]: "uniform3f",
                [Basic.vec4]: "uniform4f",
                [Basic.bvec2]: "uniform2i",
                [Basic.bvec3]: "uniform3i",
                [Basic.bvec4]: "uniform4i",
                [Basic.ivec2]: "uniform2i",
                [Basic.ivec3]: "uniform3i",
                [Basic.ivec4]: "uniform4i",
                [Basic.uvec2]: "uniform2ui",
                [Basic.uvec3]: "uniform3ui",
                [Basic.uvec4]: "uniform4ui",
                [Basic.mat2]: "uniformMatrix2fv",
                [Basic.mat3]: "uniformMatrix3fv",
                [Basic.mat4]: "uniformMatrix4fv",
                [Basic.mat2x2]: "uniformMatrix2fv",
                [Basic.mat2x3]: "uniformMatrix2fv",
                [Basic.mat2x4]: "uniformMatrix2fv",
                [Basic.mat3x2]: "uniformMatrix3fv",
                [Basic.mat3x3]: "uniformMatrix3fv",
                [Basic.mat3x4]: "uniformMatrix3fv",
                [Basic.mat4x2]: "uniformMatrix4fv",
                [Basic.mat4x3]: "uniformMatrix4fv",
                [Basic.mat4x4]: "uniformMatrix4fv",
                [Basic.sampler2D]: "uniform1i",
                [Basic.sampler3D]: "uniform1i",
                [Basic.samplerCube]: "uniform1i",
                [Basic.sampler2DArray]: "uniform1i",
                [Basic.samplerCubeShadow]: "uniform1i",
                [Basic.sampler2DShadow]: "uniform1i",
                [Basic.sampler2DArrayShadow]: "uniform1i",
                [Basic.isampler2D]: "uniform1i",
                [Basic.isampler3D]: "uniform1i",
                [Basic.isamplerCube]: "uniform1i",
                [Basic.isampler2DArray]: "uniform1i",
                [Basic.usampler2D]: "uniform1ui",
                [Basic.usampler3D]: "uniform1ui",
                [Basic.usamplerCube]: "uniform1ui",
                [Basic.usampler2DArray]: "uniform1ui",
            } as const;
        }
        export namespace Defs {
            export const Basic = {
                ...Scalar,
                ...Vector,
                ...Matrix,
                ...Sampler,
            } as const;

            interface StructName extends String {
            }

            export type Basic = keyof typeof Basic;      

            export interface Array {
                type: Basic | StructName;
                size: [number] | [number, number];
            }

            export function tagStructName<T extends string>(name: T): StructName {
                if (isStructName(name)) {
                    return name;
                }
                throw new Error(`Invalid StructName: ${name}`);
            }

            export function isStructName(type: any): type is StructName {
                return typeof type === 'string' && type.length > 0;
            }

            export function isArray(type: { type: Basic | StructName, size: Size }): type is Array {
                if (!isBasicDef(type?.type) && !isStructName(type?.type)) {
                    return false;
                }
                if (!isSize(type?.size)) {
                    return false;
                }
                return true;
            }

            export type Size = [number] | [number, number];

            export function isSize(type: any): type is Size {
                if (Array.isArray(type) && type.length === 1 && typeof type[0] === "number" && Number.isInteger(type[0]) && type[0] > 1) {
                    return true;
                }
                if (Array.isArray(type) && type.length === 2 && typeof type[0] === "number" && typeof type[1] === "number" && Number.isInteger(type[0]) && Number.isInteger(type[1]) && type[0] > 1 && type[1] > 1) {
                    return true;
                }
                return false;
            }

            export function inferSize(type: any): Size {
                if (isSize(type)) {
                    return type;
                }
                throw new Error(`Invalid Size: ${type}`);
            }

            export function inferArrayDef(type: { type: Basic | StructName, size: Size }): Array {
                if (isArray(type)) {
                    return type;
                }
                throw new Error(`Invalid Array: ${type}`);
            }
            export interface Struct {
                [key: string]: Basic | Array | StructName;
            }
        
            export function isBasicDef(type: any): type is Basic {
                return Object.values(Basic).includes(type as Basic);
            }

            export function inferBasicDef(type: any): Basic {
                if (isBasicDef(type)) {
                    return type;
                }
                throw new Error(`Invalid Basic: ${type}`);
            }

            export function inferStructNameOrBasicDef(type: any): StructName | Basic {
                if (isStructName(type)) {
                    return type;
                }
                if (isBasicDef(type)) {
                    return type;
                }
                throw new Error(`Neither StructName nor BasicDef: ${type}`);
            }

            export function isStructDef(type: any): type is Struct {
                if (typeof type === 'object' && type !== null && !Array.isArray(type)) {
                    for (const value of Object.values(type as Struct)) {
                        if (!isBasicDef(value) && !isStructName(value) && !isArray(value)) {
                            return false;
                        }
                    }
                    return true;
                }
                return false;
            }

            export function inferStructDef(type: any): Struct {
                if (isStructDef(type)) {
                    return type;
                }
                throw new Error(`Invalid StructDef: ${type}`);
            }
        }
    }
}
