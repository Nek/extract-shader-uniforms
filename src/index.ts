import { parser, generate } from "@shaderfrog/glsl-parser";
import { visit, type NodeVisitors } from "@shaderfrog/glsl-parser/ast";
import { GL, Size } from "./types";

const { tagStructName, inferSize, inferStructNameOrBasicDef } = GL.Uniform.Defs;

export function extractRelevantData(source: string) {
    let ast = parser.parse(source);
    const { bindings, types } = ast.scopes[0];
    const globalVars = Object.keys(bindings);
    const structNames = Object.keys(types);

    function getDataForTypeDef(typeSpec) {
        //TODO: implement arrays
        const [typeName, ...size] = typeSpec;
        if (structNames.includes(typeName)) {
            return structsData[typeName];
        } else {
            return typeSpec;
        }
    }

    // console.log(topLevelDefs, structs);
    // console.log(bindings, types)
    // const newSource = generate(ast);
    // ast = parser.parse(newSource);

    const uniformDefs: Record<string, any> = {};
    const uniformsData: Record<string, any> = {};
    const structDefs: Record<string, any> = {};
    const structsData: Record<string, a> = {};

    // const uniformDefs: GL.Uniform.Defs.Struct = {};
    // const uniformsData: GL.Uniform.Data.Scope = {};
    // const structsData: Record<string, GL.Uniform.Data.Scope> = {};
    // const structDefs: Record<string, GL.Uniform.Defs.Struct> = {};

    const structDefVisitors: NodeVisitors = {
        type_name: {
            enter(p) {
                const id = p.node.identifier;
                if (!structNames.includes(id)) {
                    p.skip();
                    return;
                }
                if (structDefs[id]) {
                    p.skip();
                    return;
                }
                if (p.parent?.type !== "struct") {
                    p.skip();
                    return;
                }
                let membersDefs: Record<string, any> = {};
                let membersData: Record<string, any> = {};
                for (const member of p.parent.declarations) {
                    let memberTypeName;
                    const typeSpec = member.declaration.specified_type.specifier.specifier;
                    if (typeSpec.type === "keyword") {
                        memberTypeName = typeSpec.token;
                    }
                    if (typeSpec.type === "type_name") {
                        memberTypeName = typeSpec.identifier;
                    }
                    if (!memberTypeName) {
                        break;
                    }
                    let memberId;
                    const memberSize = [];
                    for (const innerD of member.declaration.declarations) {
                        memberId = innerD.identifier.identifier;
                        if (innerD.quantifier) {
                            for (const q of innerD.quantifier) {
                                if (q.type === "array_specifier" && q.expression.type === "int_constant") {
                                    memberSize.push(parseInt(q.expression.token, 10));
                                }
                                if (memberSize.length === 2) {
                                    break;
                                }
                            }
                            if (memberSize[1] < 2) {
                                memberSize.pop();
                            }
                            if (memberSize[0] < 2) {
                                memberSize.pop();
                            }
                        }
                    }
                    const typeDef = (function (s) {
                        switch (s.length) {
                            case 1:
                                return [memberTypeName, s[0]] as [string, number];
                            case 2:
                                return [memberTypeName, s[0], s[1]] as [string, number, number];
                            default:
                                return [memberTypeName] as [string];
                        }
                    })(memberSize);
                    if (memberId) {
                        membersDefs[memberId] = typeDef;
                        membersData[memberId] = getDataForTypeDef(typeDef);
                    }
                }
                structsData[id] = membersData
                structDefs[id] = membersDefs;
            },
        },
    };

    const uniformDefVisitors: NodeVisitors = {
        identifier: {
            enter(p) {
                const id = p.node.identifier;
                if (!globalVars.includes(id)) {
                    p.skip();
                    return;
                }
                const declarator = p.parentPath?.parent;
                if (declarator?.type !== "declarator_list") {
                    p.skip();
                    return;
                }
                const keywordNode2 = declarator.specified_type.qualifiers?.[0];
                if (keywordNode2?.type !== "keyword" || keywordNode2?.token !== "uniform") {
                    p.skip();
                    return;
                }

                const typeNode = declarator.specified_type.specifier.specifier;

                const typeName = (() => {
                    if (typeNode.type === "keyword") {
                        return typeNode.token;
                    }
                    if (typeNode.type === "type_name") {
                        return typeNode.identifier;
                    }
                    return;
                })();

                if (!typeName) {
                    p.skip();
                    return;
                }

                const size = [];
                if (p.parent?.type === "declaration") {
                    const quantifier = p.parent?.quantifier;
                    if (quantifier) {
                        for (const q of quantifier) {
                            if (q.type === "array_specifier" && q.expression.type === "int_constant") {
                                size.push(parseInt(q.expression.token, 10));
                            }
                            if (size.length === 2) {
                                break;
                            }
                        }
                        if (size[1] < 2) {
                            size.pop();
                        }
                        if (size[0] < 2) {
                            size.pop();
                        }
                    }
                }
                const typeDef = (function (s) {
                    switch (s.length) {
                        case 1:
                            return [typeName, s[0]] as [string, number];
                        case 2:
                            return [typeName, s[0], s[1]] as [string, number, number];
                        default:
                            return [typeName] as [string];
                    }
                })(size);
                uniformDefs[id] = typeDef;
                uniformsData[id] = getDataForTypeDef(typeDef);
            },
        },
    };

    visit(ast, structDefVisitors);
    visit(ast, uniformDefVisitors);

    console.log(structDefs, uniformDefs);
    console.log(structsData, uniformsData);

    return {
        structDefs,
        structsData,
        uniformDefs,
        uniformsData,
    };
}
