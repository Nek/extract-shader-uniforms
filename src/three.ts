import * as THREE from 'three';
let GL = GLSL.V300ES.Type;
export const TYPES: Record<GLSL.V300ES.Type, [string, any]> = {
    [GL.Bool]: ['boolean', false],
    [GL.Int]: ['number', 0],
    [GL.Uint]: ['number', 0],
    [GL.Float]: ['number', 0],
    [GL.Vec2]: ['THREE.Vector2', new THREE.Vector2()],
    [GL.Vec3]: ['THREE.Vector3', new THREE.Vector3()],
    [GL.Vec4]: ['THREE.Vector4', new THREE.Vector4()],
    [GL.Bvec2]: ['THREE.Vector2', new THREE.Vector2()],
    [GL.Bvec3]: ['THREE.Vector3', new THREE.Vector3()],
    [GL.Bvec4]: ['THREE.Vector4', new THREE.Vector4()],
    [GL.Ivec2]: ['THREE.Vector2', new THREE.Vector2()],
    [GL.Ivec3]: ['THREE.Vector3', new THREE.Vector3()],
    [GL.Ivec4]: ['THREE.Vector4', new THREE.Vector4()],
    [GL.Uvec2]: ['THREE.Vector2', new THREE.Vector2()],
    [GL.Uvec3]: ['THREE.Vector3', new THREE.Vector3()],
    [GL.Uvec4]: ['THREE.Vector4', new THREE.Vector4()],
    [GL.Mat2]: ['THREE.Matrix2', new THREE.Matrix2()],
    [GL.Mat3]: ['THREE.Matrix3', new THREE.Matrix3()],
    [GL.Mat4]: ['THREE.Matrix4', new THREE.Matrix4()],
    [GL.Mat2x2]: ['THREE.Matrix2', new THREE.Matrix2()],
    [GL.Mat2x3]: ['[number, number, number, number, number, number]', [1, 0, 0, 1, 0, 0]],
    [GL.Mat2x4]: ['[number, number, number, number, number, number, number, number]', [1, 0, 0, 1, 0, 0, 0, 0]],
    [GL.Mat3x2]: ['[number, number, number, number, number, number]', [1, 0, 0, 1, 0, 0]],
    [GL.Mat3x3]: ['THREE.Matrix3', new THREE.Matrix3()],
    [GL.Mat3x4]: ['[number, number, number, number, number, number, number, number, number, number, number, number]', [1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]],
    [GL.Mat4x2]: ['[number, number, number, number, number, number, number, number]', [1, 0, 0, 1, 0, 0, 0, 0]],
    [GL.Mat4x3]: ['[number, number, number, number, number, number, number, number, number, number, number, number]', [1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]],
    [GL.Mat4x4]: ['THREE.Matrix4', new THREE.Matrix4()],    
    [GL.Sampler2D]: ['THREE.Texture', new THREE.Texture()],
    [GL.Sampler3D]: ['THREE.Texture', new THREE.Texture()],
    [GL.SamplerCube]: ['THREE.Texture', new THREE.Texture()],
    [GL.SamplerCubeShadow]: ['THREE.Texture', new THREE.Texture()],
    [GL.Sampler2DShadow]: ['THREE.Texture', new THREE.Texture()],
    [GL.Sampler2DArray]: ['THREE.Texture', new THREE.Texture()],
    [GL.Sampler2DArrayShadow]: ['THREE.Texture', new THREE.Texture()],
    [GL.Isampler2D]: ['THREE.Texture', new THREE.Texture()],
    [GL.Isampler3D]: ['THREE.Texture', new THREE.Texture()],
    [GL.IsamplerCube]: ['THREE.Texture', new THREE.Texture()],
    [GL.Isampler2DArray]: ['THREE.Texture', new THREE.Texture()],
    [GL.Usampler2D]: ['THREE.Texture', new THREE.Texture()],
    [GL.Usampler3D]: ['THREE.Texture', new THREE.Texture()],
    [GL.UsamplerCube]: ['THREE.Texture', new THREE.Texture()],
    [GL.Usampler2DArray]: ['THREE.Texture', new THREE.Texture()],
};
