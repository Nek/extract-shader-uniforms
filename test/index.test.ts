import { expect, test, describe } from "vitest";
import { extractRelevantData } from "../src/index";
import { GL } from "../src/types";
describe("extractRelevantData", () => {
    test("should extract primitive types from a shader", () => {
        const source = `
            uniform float u_time;
            uniform vec2 u_resolution;
            uniform vec3 u_color[2][2];
            uniform vec4 u_position;
            float testVar;
        `;
        const result = extractRelevantData(source);
        expect(result).toMatchSnapshot();
    });
    test("should extract structs from a shader", () => {
        let source = `
        
        uniform sampler2D uText;
        struct Light {
            vec3 position;
            vec3 color;
            };
        uniform Light u_light[2];
  `;
        const result = extractRelevantData(source);
        expect(result).toMatchSnapshot();
    });
    test("should extract structs from a shader", () => {
        const source = `
        struct Light {
            vec3 position;
            vec3 color;
        };
        uniform Light u_light[3];
  `;
        const nestedLight = {
            color: GL.Uniform.Data.inferBasicData("vec3"),
            position: GL.Uniform.Data.inferBasicData("vec3"),
        };
        const expectedUniformsData: GL.Uniform.Data.Scope = {
            u_light: [
                nestedLight,
                nestedLight,
                nestedLight,
            ],
        };
        const result = extractRelevantData(source);
        expect(result.uniformsData).toEqual(expectedUniformsData);
    });
    test("should extract arrays from a shader", () => {
        const source = `
        uniform vec2 u_resolution[2];
        `;
        const result = extractRelevantData(source);
        expect(result.uniformsData).toEqual({
            u_resolution: ["vec2", "vec2"
            ],
        });
    });
    test("should extract structs from a shader", () => {
        const source = `
        struct Camera {
            vec3 position;
            vec3 color;
        };
        struct Light {
            Camera camera;
            vec3 position[2];
            vec3 color;
        };
        uniform Light u_light[2];
        `;
        const result = extractRelevantData(source);

        console.log(JSON.stringify(result, null, 4))

        expect(result.uniformsData).toEqual({
            u_light: [
                {
                    camera: {
                        color: "vec3",
                        position: "vec3",
                    },
                    color: "vec3",
                    position: ["vec3", "vec3"],
                },
                {
                    color: "vec3",
                    position: ["vec3", "vec3"],
                    camera: {
                        color: "vec3",
                        position: "vec3",
                    },
                },
            ],
        });
    });
});
