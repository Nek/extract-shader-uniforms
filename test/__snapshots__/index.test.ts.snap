// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractRelevantData > should extract primitive types from a shader 1`] = `
{
  "structDefs": {},
  "structsData": {},
  "uniformDefs": {
    "u_color": "vec3",
    "u_position": "vec4",
    "u_resolution": "vec2",
    "u_time": "float",
  },
  "uniformsData": {
    "u_color": "vec3",
    "u_position": "vec4",
    "u_resolution": "vec2",
    "u_time": "float",
  },
}
`;

exports[`extractRelevantData > should extract structs from a shader 1`] = `
{
  "structDefs": {
    "Light": {
      "color": "vec3",
      "position": "vec3",
    },
  },
  "structsData": {
    "Light": {
      "color": "vec3",
      "position": "vec3",
    },
  },
  "uniformDefs": {
    "u_light": {
      "size": [
        2,
      ],
      "type": "Light",
    },
  },
  "uniformsData": {
    "u_light": [
      {
        "color": "vec3",
        "position": "vec3",
      },
      {
        "color": "vec3",
        "position": "vec3",
      },
    ],
  },
}
`;

exports[`extractRelevantData should extract primitive types from a shader 1`] = `
{
  "structDefs": {},
  "structsData": {},
  "uniformDefs": {
    "u_color": "vec3",
    "u_position": "vec4",
    "u_resolution": "vec2",
    "u_time": "float",
  },
  "uniformsData": {
    "u_color": "vec3",
    "u_position": "vec4",
    "u_resolution": "vec2",
    "u_time": "float",
  },
}
`;

exports[`extractRelevantData should extract structs from a shader 1`] = `
{
  "structDefs": {
    "Light": {
      "color": "vec3",
      "position": "vec3",
    },
  },
  "structsData": {
    "Light": {
      "color": "vec3",
      "position": "vec3",
    },
  },
  "uniformDefs": {
    "u_light": {
      "size": [
        2,
      ],
      "type": "Light",
    },
  },
  "uniformsData": {
    "u_light": [
      {
        "color": "vec3",
        "position": "vec3",
      },
      {
        "color": "vec3",
        "position": "vec3",
      },
    ],
  },
}
`;
